presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java

```diff
@@ -1690,16 +1726,59 @@ public void commit()
         metastore.commit();
     }
 
-    public static boolean isPartitionsSystemTable(SchemaTableName tableName)
+    public static Optional<SchemaTableName> getSourceTableNameFromSystemTable(SchemaTableName tableName)
     {
-        return tableName.getTableName().endsWith(PARTITIONS_TABLE_SUFFIX) && tableName.getTableName().length() > PARTITIONS_TABLE_SUFFIX.length();
+        for (SystemTableHandler handler : ImmutableList.of(partitionsTableHandler, propertiesTableHandler)) {
+            if (handler.matches(tableName)) {
+                SchemaTableName sourceTableName = handler.getSourceTableName(tableName);
+                return Optional.of(sourceTableName);
+            }
+        }
+        return Optional.empty();
     }
 
-    public static SchemaTableName getSourceTableNameForPartitionsTable(SchemaTableName tableName)
+    interface SystemTableHandler
     {
-        checkArgument(isPartitionsSystemTable(tableName), "not a partitions table name");
-        return new SchemaTableName(
-                tableName.getSchemaName(),
-                tableName.getTableName().substring(0, tableName.getTableName().length() - PARTITIONS_TABLE_SUFFIX.length()));
+        boolean matches(SchemaTableName schemaTableName);
+
+        SchemaTableName getSourceTableName(SchemaTableName tableName);
     }
+
+    private static SystemTableHandler partitionsTableHandler = new SystemTableHandler()
```

Comment:
The code for both of these is the same, except for the suffix. I think we could eliminate some redundancy:
```java
private interface SystemTableHandler
{
    String getSuffix();

    boolean matches(SchemaTableName table)
    {
        return table.getTableName().endsWith(getSuffix()) && 
                (table.getTableName().length() > getSuffix().length());
    }

    SchemaTableName getSourceTableName(SchemaTableName table)
    {
        return new SchemaTableName(
                table.getSchemaName(),
                table.getTableName().substring(table.getTableName().length() - getSuffix().length()));
    }
}

private static final SystemTableHandler PARTITIONS_TABLE_HANDLER = () -> "$partitions";
private static final SystemTableHandler PROPERTIES_TABLE_HANDLER = () -> "$properties";